src/domain/proposal.ts

```diff
@@ -0,0 +1,145 @@
+import { Address, BigInt, Bytes, crypto, store } from '@graphprotocol/graph-ts';
+import { GenesisProtocol } from '../types/GenesisProtocol/GenesisProtocol';
+import { Proposal } from '../types/schema';
+import { concat, equals } from '../utils';
+import { getMember } from './member';
+
+export function parseOutcome(num: BigInt): string {
+  if (equals(num, BigInt.fromI32(1))) {
+    // Yes
+    return 'Pass';
+  } else {
+    // No
+    return 'Fail';
+  }
+}
+
+export function getProposal(id: string): Proposal {
+  let proposal = store.get('Proposal', id) as Proposal;
+  if (proposal == null) {
+    proposal = new Proposal();
+    proposal.id = id;
+
+    proposal.stage = 'Open';
+
+    proposal.votesFor = BigInt.fromI32(0);
+    proposal.votesAgainst = BigInt.fromI32(0);
+    proposal.winningOutcome = 'Fail';
+
+    proposal.stakesFor = BigInt.fromI32(0);
+    proposal.stakesAgainst = BigInt.fromI32(0);
+  }
+
+  return proposal;
+}
+
+export function saveProposal(proposal: Proposal): void {
+  store.set('Proposal', proposal.id, proposal);
+}
+
+export function updateProposal(
+  proposal: Proposal,
+  gpAddress: Address,
+  proposalId: Bytes,
+): void {
+  let gp = GenesisProtocol.bind(gpAddress);
+  let gpProposal = gp.proposals(proposalId);
+
+  // proposal.boostedPhaseTime
+  if (!equals(gpProposal.value5, BigInt.fromI32(0))) {
+    if (proposal.boostedAt == null) {
+      proposal.boostedAt = gpProposal.value5;
+    } else if (!equals(proposal.boostedAt as BigInt, gpProposal.value5)) {
+      proposal.overtimedAt = gpProposal.value5;
+    }
+  }
+
+  // proposal.winningVote
+  proposal.winningOutcome = parseOutcome(gpProposal.value7);
+
+  // proposal.state
+  let state = gpProposal.value6;
+  if (state === 1) {
+    // Closed
+    proposal.stage = 'Resolved';
```

Comment:
'Resoved' == 'Closed' ?

Reply:
Resolved == Executed || Closed. To check whether it is executed you use `executedAt != null`.

Reply:
how about differentiate here between Executed and Closed (Rejected) ?

Reply:
1. `executed == Resolved && executedAt != null`
2. `closed == Resolved && executedAt == null`

Reply:
what does executedAt means ?